packages/material-ui/src/Button/Button.js

```diff
@@ -58,19 +58,19 @@ const ButtonRoot = styled(ButtonBase, {
   shouldForwardProp: (prop) => rootShouldForwardProp(prop) || prop === 'classes',
   name: 'MuiButton',
   slot: 'Root',
-  overridesResolver: (props, styles) => {
+  overridesResolver: (props) => {
     const { styleProps } = props;
 
-    return {
-      ...styles.root,
-      ...styles[styleProps.variant],
-      ...styles[`${styleProps.variant}${capitalize(styleProps.color)}`],
-      ...styles[`size${capitalize(styleProps.size)}`],
-      ...styles[`${styleProps.variant}Size${capitalize(styleProps.size)}`],
-      ...(styleProps.color === 'inherit' && styles.colorInherit),
-      ...(styleProps.disableElevation && styles.disableElevation),
-      ...(styleProps.fullWidth && styles.fullWidth),
-    };
+    return [
+      'root',
+      styleProps.variant,
+      `${styleProps.variant}${capitalize(styleProps.color)}`,
+      `size${capitalize(styleProps.size)}`,
+      `${styleProps.variant}Size${capitalize(styleProps.size)}`,
+      styleProps.color === 'inherit' && 'colorInherit',
+      styleProps.disableElevation && 'disableElevation',
+      styleProps.fullWidth && 'fullWidth',
+    ];
```

Comment:
@oliviertassinari one thing that I concern about is performance before the latter needs to call `.map` again.

**Approach I**
similar to current but just return array instead of object.
```js
[
  styles.root,
  styles[styleProps.variant],
  ...
]
```

**Approach II**
return array of classKey and createStyled will iterate to generate styles.
```js
[
 'root',
  styleProps.variant,
  ...
]
```
@eps1lon any thought on the performance. For context, the PR is trying to fix #26813. The solution is to return array instead of object so that the styles at the bottom does not override the top. (I tested with emotion & styled-components, both support array)

Reply:
@siriwatknp No real preferences. I also don't expect a difference in performance, maybe option 1 is better to be more explicit with little to no boilerplate downside?

Reply:
@oliviertassinari Yeah, I think let's move forward with option 1, I is straight forward with nested classes.

```jsx
// ButtonGroup.js
const overridesResolver = (props, styles) => {
  const { styleProps } = props;

  return {
    [`& .${buttonGroupClasses.grouped}`]: {
      ...styles.grouped,
      ...styles[`grouped${capitalize(styleProps.orientation)}`],
      ...styles[`grouped${capitalize(styleProps.variant)}`],
      ...styles[`grouped${capitalize(styleProps.variant)}${capitalize(styleProps.orientation)}`],
      ...styles[`grouped${capitalize(styleProps.variant)}${capitalize(styleProps.color)}`],
    },
    ...styles.root,
    // ...other styles
  };
};
```